README.md
[![Latest Stable Version](https://poser.pugx.org/andrebian/pag_seguro/v/stable.png)](https://packagist.org/packages/andrebian/pag_seguro) [![Total Downloads](https://poser.pugx.org/andrebian/pag_seguro/downloads.png)](https://packagist.org/packages/andrebian/pag_seguro) [![Latest Unstable Version](https://poser.pugx.org/andrebian/pag_seguro/v/unstable.png)](https://packagist.org/packages/andrebian/pag_seguro) [![License](https://poser.pugx.org/andrebian/pag_seguro/license.png)](https://packagist.org/packages/andrebian/pag_seguro)
[![githalytics.com alpha](https://cruel-carlota.pagodabox.com/8d4bc51766116b27f682121865660505 "githalytics.com")](http://githalytics.com/andrebian/cake-plugin-pagseguro)

# PAGSEGURO PLUGIN
_v 2.1.2_


Facilita a integração de pagamentos via PagSeguro em aplicações desenvolvidas com base no CakePHP 2.x.
O plugin realiza apenas interfaceamento para a API de pagamentos do PagSeguro, com
isso nem o plugin nem o PagSeguro podem ser responsabilizados por uso em desconformidade à documentação fornecida pelo PagSeguro <https://pagseguro.uol.com.br/v2/guia-de-integracao/visao-geral.html> assim como valores fornecidos. A responsabilidade das corretas informações ao PagSeguro são estritamente do programador que criará a requisição no fechamento do carrinho de compras.

____________________

## INSTALAÇÃO

Composer
---------------
```json
{
    "require": {
        "andrebian/pag_seguro": "dev-master"
    }
}
```
Ou se preferir uma versão em específico:

```json
{
    "require": {
        "andrebian/pag_seguro": "2.0.0"
    }
}
```

Ainda

```json
{
    "require": {
        "andrebian/pag_seguro": "*"
    }
}
```

##Erro comum

Por depender da versão de desenvolvimento do PagSeguro pode ser necessário marcar a estabilidade mínima dos pacotes 
a serem instalados.

```json
{
    "require": {
        "andrebian/pag_seguro": "*"
    },
    "minimum-stability": "dev"
}
```


##Importantíssimo

Por padrão o CakePHP 2.x utiliza plugins dentro de `app/Plugin/` mas ao instalar diretamente 
pelo composer o mesmo será instalado na raíz de sua aplicação (Plugin/) e não em app/Plugin/. 
Para contornar isto existem duas saídas, a primeira é definir diretamente através do seu arquivo 
`composer.json` e a segunda é informar o novo ponto de acesso à plugins. Ambos serão descritos a seguir.

###Opção 1 - Definindo pelo composer.json

Em seu arquivo`composer.json` informe o seguinte:

```json
"extra": {
    "installer-paths": {
        "app/Plugin/PagSeguro": ["andrebian/pag_seguro"]
    }
}
```

###Opção 2 - Informando no bootstrap.php 

É possível informar no bootstrap.php um novo ponto de acesso à plugins, desta forma o Cake 2.x 
procurará os plugins nos seguintes endereços: `Plugin/` a `app/Plugin/`.

```php
App::build(array('Plugin' => array(ROOT . '/Plugin/', ROOT . '/app/Plugin/')));
```

> Ao seguir esta opção a primeira pode ser ignorada.


Para mais informações sobre estas opções leia este post em meu blog e veja ao final que 
menciono uma issue do Debug Kit onde o @markstory, @ADmad e @ceeram explicam os motivos e estrutura do Cake e sugerem a opção do bootstrap.php: http://www.andrebian.com/cakephp-e-composer/ 

##Atenção
> Não é mais fornecido suporte para download direto ou utilização como submodulo. Por depender da API oficial do PagSeguro a instalação faz-se de forma correta através do composer. Se quiser fazer desta forma (clone, zip ou submodulo) faça, mas não prestarei suporte pois o único meio que satisfaz todas as dependências é o indicado acima.

_________________________

## CONFIGURAÇÕES


### Carregando o plugin

No arquivo `bootstrap.php` adicione o suporte ao plugin:
`CakePlugin::load('PagSeguro');` ou `CakePlugin::loadAll();`


### Credenciais

Você deve possuir uma conta no PagSeguro pois precisará setar as credenciais,
estas credenciais são compostas pelo seu email e o token que deve ser configurado na seção de integração
junto ao PagSeguro.

Tal configuração pode ser feita de duas formas, via `bootstrap` ou no controller desejado.

Arquivo bootstrap
```php
Configure::write('PagSeguro', array(
    'email' => 'seu-email-cadastrado@pagseguro',
    'token' => 'seu-token',
    'isSandbox' => true, // true|false
));
```        


Controller qualquer onde será montada a finalização da compra
```php
$this->Checkout->defineCredenciais('email cadastrado', 'token gerado');
```


### Carregando o componente


Agora que você já configurou suas credenciais deve definir no `AppController` ou no controller
que o componente será utilizado

```php
public $components = array('PagSeguro.PagSeguro', 'PagSeguro.Checkout', 'PagSeguro.RetornoPagSeguro');
```


Caso já possua mais componentes faça-o da seguinte forma

```php
public $components = array('Outros componentes', 'PagSeguro.PagSeguro', 'PagSeguro.Checkout', 'PagSeguro.RetornoPagSeguro');
```


**A utilização do `'PagSeguro.PagSeguro'` NÃO deve ser ignorada.**


## UTILIZAÇÃO


* Requisição de pagamento: Leia o arquivo [REQUISICAO_PAGAMENTO][1]
* Retorno de requisição de pagamento: Leia o arquivo [CONSULTA][2]
* Consulta por código: Leia o arquivo [CONSULTA][3]
* Consulta por perído: Leia o arquivo [CONSULTA][4]
* Consulta por transações abandonadas: Leia o arquivo [CONSULTA][5]
* Notificações: Leia o arquivo [NOTIFICACAO][6]

______________


##TODO

* Testes


##PROBLEMAS

Reporte-os por aqui https://github.com/andrebian/cake-plugin-pagseguro/issues?state=open

##CONTRIBUIDORES

* Marcelo Siqueira - https://github.com/marcelosiqueira


##ATENÇÃO
Muitos desenvolvedores utilizaram meu email pessoal que estava nos exemplos para realizar compras, peço humildemente que por favor, use SEU email para realizar os testes de compra, já ajudei com o plugin não posso ajudar verificando emails e dando reports.

Muito agradecido.


  [1]: https://github.com/andrebian/cake-plugin-pagseguro/blob/master/REQUISICAO_PAGAMENTO.md
  [2]: https://github.com/andrebian/cake-plugin-pagseguro/blob/master/CONSULTA.md
  [3]: https://github.com/andrebian/cake-plugin-pagseguro/blob/master/CONSULTA.md
  [4]: https://github.com/andrebian/cake-plugin-pagseguro/blob/master/CONSULTA.md
  [5]: https://github.com/andrebian/cake-plugin-pagseguro/blob/master/CONSULTA.md
  [6]: https://github.com/andrebian/cake-plugin-pagseguro/blob/master/NOTIFICACAO.md


CHANGELOG
################################################################################
#                   PAGSEGURO PLUGIN CAKEPHP 2.x                               #
#                                                                              #
#                                                                              #
#                       LOG DE ALTERAÇÕES                                      #
################################################################################

2.1.2 - 16/01/2015
    * Corrigidos includes de componentes antigos mesmo que em deprecated
    * Adicionada como dependência a versão de desenvolvimeto do PagSeguro para que a notificação funcionasse corretamente

2.1.1
    * Realizadas algumas correções pós ambiente sandbox

2.1 - 02/01/2015
    * Implementado ambiente sandbox
    * Novos componentes em pt_BR
    * componentes antigs marcados como deprecated


2.0.0 - 14/05/2014
    * Removida biblioteca do PagSeguro de Vendor - Agora a mesma é carregada com o composer
    * Alterado local da classe de Códigos de status e pagamentos
    * Adicionada possibilidade de definir a URL de notificação da compra

1.5.1 - 28/09/2013
    * Corrigido retorno de peso de itens adquiridos. Anteriormente era a descrição do
        item que era trazida ao invés do peso

1.5 - 02/07/2013
    * Adicionado retorno do código da transação ao receber notificação

1.4.1 - 30/04/2013
    * Adicionada referência de compra na notificação

1.4.1 - 03/04/2013
    * Adicionada referência da compra na consulta

1.4.1 - 18/03/2013
    * Alterada codificação padrão do XML parser para Utf8

1.4 - 26/02/2013
    * Criada possibilidade de consulta por transações abandonadas

1.3 - 23/02/2013
    * Criada possibilidade de consulta por período
    * Documentação separada por módulos

1.2 - 23/02/2013
    Adicionados tipo de entrega, valor do frete e tipo de pagamento


1.1 - 22/02/2013
    * Criados retorno de dados de produtos comprados em carrinho e notificações
    * Melhorada documentação - adicionados exemplos de retornos
    * Definido tipo de licença MIT (http://opensource.org/licenses/MIT)
    * Adicionada versão da API do PagSeguro
    

1.0 - 13/02/2013 
    Primeira versão do plugin


Controller/Component/CarrinhoComponent.php
<?php
/**
* Plugin de integração com a API do PagSeguro e CakePHP (API de verdade, não somente montar a string e enviar por post).
*
* Versão do PHP 5.*
*
*
* @author Andre Cardoso <andrecardosodev@gmail.com>
* @link https://github.com/andrebian/cake-plugin-pagseguro/
* @authorURI http://andrebian.com
* @license MIT (http://opensource.org/licenses/MIT)
* @version 2.0.0
* @since 1.0 
* 
* ESTE PLUGIN UTILIZA A API DO PAGSEGURO, DISPONÍVEL EM  https://pagseguro.uol.com.br/v2/guia-de-integracao/tutorial-da-biblioteca-pagseguro-em-php.html
* 
* 
* 
* PLUGIN BASE (enviando dados via POST):
*   https://github.com/ftgoncalves/pagseguro/  de Felipe Theodoro Gonçalves, (http://ftgoncalves.com.br)
*/

class CarrinhoComponent extends PagSeguroComponent
{

    private $montaPagamento = null;
    private $consultaPorCodigo = null;
    private $comprador = null;
    private $tipoPagamento = null;
    
   
/**
 * 
 * @param \Controller $Controller
 * @since 1.0
 */    
    public function startup(\Controller $Controller) 
    {
        parent::startup($Controller);

        // Instanciando classes para gerar o pagamento
        $this->montaPagamento = new PagSeguroPaymentRequest;
        $this->comprador = new PagSeguroSender;
        $this->tipoPagamento = new PagSeguroPaymentMethodType;
        
        $this->montaPagamento->setShippingType('3');
        $this->montaPagamento->setCurrency('BRL');
    }
  
    
 /**
  * @deprecated since version 2.0.3
  * 
  * Use CheckoutComponent::defineCredenciais para novas chamadas
  * 
  * Define as credenciais para utilização do PagSeguro
  *
  * @param string $email
  * @param string $token
  * @since 1.0
  */   
    public function setCredenciais($email, $token) {
        $this->credenciais = new PagSeguroAccountCredentials($email, $token);
    }
    
    /**
     * @deprecated since version 2.0.3
     * 
     * Não mais presente a partir da versão 2.1
     * 
     * @param string $url
     */
    public function setUrlNotificacoes( $url = null )
    {
        $this->montaPagamento->setNotificationURL($url);
    }
    
    
 /**
  * @deprecated since version 2.0.3
  * 
  * Use CheckoutComponent::defineUrlNotificacoes
  * 
  * Define uma URL para após a finalização da interação do usuário com o PagSeguro
  * 
  * @param string $urlRetorno
  * @since 1.0
  */   
    public function setUrlRetorno($urlRetorno = null) {
       $this->montaPagamento->setRedirectURL($urlRetorno); 
    }
    
    
 /**
  * @deprecated since version 2.0.3
  * 
  * Use CheckoutComponent::defineReferencia nas próximas chamadas
  * 
  * Seta a referência para a compra
  * 
  * @param int $id
  * @since 1.0
  */
    public function setReferencia($id) {
        $this->montaPagamento->setReference($id);
    }
    
    
/**
 * @deprecated since version 2.0.3
 * 
 * Use CheckoutComponent::adicionarItem nas próximas chamadas
 * 
 * Adiciona um único item de cada vez ao "carrinho de compras"
 * 
 * @param int $id
 * @param string $nomeProduto
 * @param float $valorUnit
 * @param int $peso
 * @param int $quantidade
 * @param float $frete
 * @since 1.0
 */
    public function adicionarItem($id, $nomeProduto, $valorUnit, $peso, $quantidade = 1, $frete = null) {
        $this->montaPagamento->addItem($id, $nomeProduto, $quantidade, $valorUnit, $peso, $frete);
    }
    
    
 /**
  * @deprecated since version 2.0.3
  * 
  * Use CheckoutComponent::defineValorExtra nas próximas chamadas
  * 
  * Define um valor extra, ponto flutuante com 2 casas decimais
  * Pode ser valor positivo para acréscimo ou negativo para desconto
  * 
  * @param string $valor (formato 0.00)
  * @since 1.0
  */   
    public function setValorExtra($valor) {
        $this->montaPagamento->setExtraAmount($valor);
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::defineValidadeRequisicao nas próximas chamadas
  * 
  * Define em segundos por quanto tempo a requisição será válida
  * 
  * @param int $validade
  * @since 1.0
  */   
    public function setValidadeRequisicao($validade) {
        $this->montaPagamento->setMaxAge($validade);
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::defineQuantidadeUso nas próximas chamadas
  * 
  * Define a quantidade de vezes que a requisição será utilizada
  * Útil para produtos ou taxas não variáveis
  * 
  * @param int $quantidade
  * @since 1.0
  */   
    public function setQuantidadeUso($quantidade) {
        $this->montaPagamento->setMaxUses($quantidade);
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::defineContatosComprador nas próximas chamadas
  * 
  * Define os dados de contato do comprador
  * 
  * @param string $nome
  * @param string $email
  * @param string $codigoArea
  * @param string $numeroTelefone
  * @since 1.0
  */   
    public function setContatosComprador($nome, $email, $codigoArea, $numeroTelefone) {
        $this->montaPagamento->setSender($nome, $email, $codigoArea, $numeroTelefone);
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::defineEnderecoComprador nas próximas chamadas
  * 
  * Define o endereço do comprador
  * 
  * @param string $cep
  * @param string $rua
  * @param string $numero
  * @param string $complemento
  * @param string $bairro
  * @param string $cidade
  * @param string $uf
  * @param string $pais Padrão BRA
  * @since 1.0
  */   
    public function setEnderecoComprador($cep, $rua, $numero, $complemento, $bairro, $cidade, $uf, $pais = 'BRA') {
        $this->montaPagamento->setShippingAddress($cep, $rua, $numero, $complemento, $bairro, $cidade, $uf, $pais);
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::defineTipoFrete nas próximas chamadas
  * 
  * Define o tipo de frete que será efetuado na compra
  * 
  * @param string $tipoFrete 
  * 1 	Encomenda normal (PAC).
  * 2 	SEDEX (SEDEX)
  * 3 	Tipo de frete não especificado (NAO_ESPECIFICADO ou outro qualquer) -> Padrão
  * @since 1.2
  * 
  */   
    public function setTipoFrete($tipoFrete) {
        
        switch ($tipoFrete) {
            case 'PAC':
                $tipoEntrega = '1';
                break;
            case 'SEDEX':
                $tipoEntrega = '2';
                break;
            default:
                $tipoEntrega = '3';
                break;
        }
        
        $this->montaPagamento->setShippingType($tipoEntrega);
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::defineValorTotalFrete nas próximas chamadas
  * 
  * Define o valor total do frete na compra
  * 
  * @param string $valorTotalFrete
  * @since 1.2
  */   
    public function setValorTotalFrete($valorTotalFrete) {
        $this->montaPagamento->setShippingCost($valorTotalFrete);
    }
    
    
/**
 * @deprecated since version 2.0.3.3
 * 
 * Use CheckoutComponent::defineTipoPagamento nas próximas chamadas
 * 
 * Define o tipo de pagamento que será disponibilizado ao cliente.
 * Detalhe: O tipo de pagamento condiz somente com os tipos habilitados em sua conta
 * junto ao PagSeguro
 * 
 * @param string tipoPagamento (CREDIT_CARD, BOLETO, ONLINE_TRANSFER, BALANCE, OI_PAGGO)
 * @since 1.2
 * 
 */    
    public function setTipoPagamento($tipoPagamento) {
        $this->tipoPagamento->setByType($tipoPagamento);
    }    
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use CheckoutComponent::finalizaCompra nas próximas chamadas
  * 
  * E enfim enviando o usuário para o pagamento
  * 
  * @return boolean
  * @since 1.0
  */   
    public function finalizaCompra() {
        try {
            if ($url = $this->montaPagamento->register($this->credenciais) ) {
                return $url;
            }
        } catch (PagSeguroServiceException $e) {
            echo $e->getMessage();
            exit();
        }
    }
    
    
    ###############
    # MÓDULO PARA TRATAR DO RETORNO
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use RetornoPagSeguroComponent::obterInformacoesTransacoes nas próximas chamadas
  * 
  * Realiza a busca através do transaction id vindo do pagseguro via get ou post 
  * em sua URL de retorno
  * 
  * @param string $idTransacao
  * @return object
  * @since 1.0
  */    
    public function obterInformacoesTransacao($idTransacao) {
        try{
            if ($this->consultaPorCodigo = PagSeguroTransactionSearchService::searchByCode($this->credenciais, $idTransacao) ) {
                return true;
            }
        } catch (PagSeguroServiceException $e) {
            echo $e->getMessage();
            exit();
        }
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use RetornoPagSeguroComponent::obterDadosUsuario nas próximas chamadas
  * 
  * Retorna os dados do usuário após a consulta
  * 
  * @return array
  * @since 1.0
  */   
    public function obterDadosUsuario() {
        $contato = $this->consultaPorCodigo->getSender();
        $endereco = $this->consultaPorCodigo->getShipping()->getAddress();
        
        $dadosUsuario = 
            array(  'nome' => $contato->getName(),
                    'email' => $contato->getEmail(),
                    'telefoneCompleto' => $contato->getPhone()->getAreaCode().' '.$contato->getPhone()->getNumber(),
                    'codigoArea' => $contato->getPhone()->getAreaCode(),
                    'numeroTelefone' => $contato->getPhone()->getNumber(),
                    'endereco' => $endereco->getStreet(),
                    'numero' => $endereco->getNumber(),
                    'complemento' => $endereco->getComplement(),
                    'bairro' => $endereco->getDistrict(),
                    'cidade' => $endereco->getCity(),
                    'cep' => $endereco->getPostalCode(),
                    'uf' => $endereco->getState(),
                    'pais' => $endereco->getCountry()
             ) ;
        
        return $dadosUsuario;
    }

 
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use RetornoPagSeguroComponent::obterStatusTransacao nas próximas chamadas
  * 
  * Retorna em modo de array o status da transação pesquisada
  * 
  * @return array
  * @since 1.0
  */   
    public function obterStatusTransacao() {
        return Codes::obterStatusTransacao($this->consultaPorCodigo->getStatus()->getValue(), array());
    }
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use RetornoPagSeguroComponent::obterReferencia nas próximas chamadas
  * 
  * Retorna em modo de array o status da transação pesquisada
  * 
  * @return array
  * @since 1.0
  */   
    public function obterReferencia() {
        return $this->consultaPorCodigo->getReference();
    }    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use RetornoPagSeguroComponent::obterDadosPagamento nas próximas chamadas
  * 
  * Retorna tipo e meio de pagamento
  * 
  * @return array
  * @since 1.0
  */   
    public function obterDadosPagamento() {
        $dadosPagamento = array(
            'tipo_id' => $this->consultaPorCodigo->getPaymentMethod()->getType()->getValue(),
            'tipo' => Codes::obterTipoPagamento($this->consultaPorCodigo->getPaymentMethod()->getType()->getValue()),
            'metodo_id' => $this->consultaPorCodigo->getPaymentMethod()->getCode()->getValue(),
            'metodo' => Codes::obterMeioPagamento($this->consultaPorCodigo->getPaymentMethod()->getCode()->getValue())
        );
        
        return $dadosPagamento;
    }
    
    
 /**
  * @deprecated since version 2.0.3.3
  * 
  * Use RetornoPagSeguroComponent::obterDataTransacao nas próximas chamadas
  * 
  * Retorna um array contendo a data em forma iso para o banco de dados
  * e em ptBR para exibição somente
  * 
  * @return array
  * @since 1.0
  */   
    public function obterDataTransacao() {
        $data['iso'] = $this->consultaPorCodigo->getDate();
        $data['ptBr'] = date('d/m/Y H:i:s', strtotime($data['iso']));
        $data['ultimaTentativaIso'] = $this->consultaPorCodigo->getLastEventDate();
        $data['ultimaTentativaPtBr'] = date('d/m/Y H:i:s', strtotime($data['ultimaTentativaIso']));
        
        return $data;
    }
    
    
   /**
    * @deprecated since version 2.0.3.3
    * 
    * Use RetornoPagSeguroComponent::obterValores nas próximas chamadas
    * 
    * obtervalores method
    * 
    * @return array
    * @since 1.1
    */   
    public function obterValores() {
        foreach($this->consultaPorCodigo->getItems() as $item) {
          $itens[] = array(
            'id' => $item->getId(), 
            'descricao' => $item->getDescription() , 
            'quantidade' => $item->getQuantity(),
            'valorUnitario' => $item->getAmount(),
            'peso' => $item->getWeight(),
            'frete' => $item->getShippingCost(),

          );
        }
        
        $dados = array(
          'referencia' => $this->consultaPorCodigo->getReference(),
          'valorTotal' => $this->consultaPorCodigo->getGrossAmount(),
          'descontoAplicado' => $this->consultaPorCodigo->getDiscountAmount(),
          'valorExtra' => $this->consultaPorCodigo->getExtraAmount(),
          'valorTaxa' => $this->consultaPorCodigo->getFeeAmount(),
          'produtos' => $itens,
        );
        
        return $dados;
    }
    
    
}